### PR TITLE
Next round of scatter-gather

### PR DESCRIPTION
**Pull request by scottfrazer**

### CODE REVIEW

**Comment by Horneth** [comment]
I think we're good, I got this wdl passing which I think cover all cases we should support for now :
```
task B {
  String B_in
  command {
    python -c "print('${B_in}')"
  }
  output {
    String B_out = read_string(stdout())
  }
}

task D {
  String D_in
  command {
    python -c "print(len('${D_in}'))"
  }
  output {
    Int D_out = read_int(stdout())
  }
}

task C {
  Array[String] C_in
  command {
    echo ${sep = ' ' C_in}
  }
  output {
    String C_out = read_string(stdout())
  }
}

workflow w {
 Array[String] an_array = ["a", "ab", "abc"]

 scatter (item in an_array) {
   call B {input: B_in = item }
   call D {input: D_in = B.B_out }
 }

 call C {input: C_in = B.B_out }
 call C as E {input: C_in = D.D_out }

 scatter (item in D.D_out) {
   call B as F {input: B_in = item }
 }
}
```
Output:
```
{
  "w.F.B_out": ["1", "2", "3"],
  "w.C.C_out": "a ab abc",
  "w.E.C_out": "1 2 3",
  "w.D.D_out": [1, 2, 3],
  "w.B.B_out": ["a", "ab", "abc"]
}
```

**Comment by mcovarr** [comment]
Excellent!  The build failures are due to the MySQL quirks addressed by #160; once that's merged to scatter-gather and this is rebased I expect the build will go green.

**Comment by scottfrazer** [comment]
@Horneth Awesome!!

File: src/main/scala/cromwell/engine/backend/BackendCall.scala [hunk 1]
```diff
@@ -45,7 +46,8 @@ trait BackendCall {
    * of a BackendCall object that the 'call' would be within the workflow
    */
   def workflowDescriptor: WorkflowDescriptor
-  def call: Call
+  def key: CallKey
+  val call = key.scope
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/BackendCall.scala]: Things in traits should be def unless there's a very good reason not to

File: src/main/scala/cromwell/engine/backend/local/LocalBackend.scala [hunk 2]
```diff
@@ -40,11 +42,19 @@ object LocalBackend {
   def hostExecutionPath(workflowName: String, workflowUuid: WorkflowId): Path =
     Paths.get(CromwellExecutions, workflowName, workflowUuid.id.toString)
 
-  def hostCallPath(workflow: WorkflowDescriptor, callName: String): Path =
-    Paths.get(hostExecutionPath(workflow).toFile.getAbsolutePath, s"call-$callName")
+  def hostCallPath(workflow: WorkflowDescriptor, callName: String, callIndex: ExecutionIndex): Path = {
+   val rootCallPath = Paths.get(hostExecutionPath(workflow).toFile.getAbsolutePath, s"call-$callName")
+    callIndex map { index =>
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/local/LocalBackend.scala]: I think if a map/getOrElse is going to span three lines anyways you might as well go with a pattern match

**Comment by geoffjentry** [comment]
Are all of the new `Await.result` calls intended to stick around or are they just tied to this being in progress?

**Comment by scottfrazer** [comment]
Regarding the `Await.Result`: Yes, we both knew this was sub optimal.  The reason it's that way is because I wanted to do something like this, for testing purposes:

```
printExecutionTable() // queries database.
Await.result(...)
Await.result(...)
printExecutionTable()
```

And when I used `Await.result`, this worked as expected: the second `printExecutionTable()` printed the new entries.  However, when I switched this to be a for comprehension, the behavior changed:  The second `printExecutionTable()` printed the same thing as the first one.

Just because I wanted to keep moving in getting this implemented, I made a mental note of it but then left the `Await.result`s in there

**Comment by geoffjentry** [comment]
@scottfrazer Cool, thanks for the explanation.

**Comment by scottfrazer** [comment]
I ran the WDL mentioned here, as well as the one I concocted a while ago: https://gist.github.com/scottfrazer/e3c4e7f804448293e714.

This one ran perfectly fine first time!  I was blown away, I was bracing myself for *something* to fail, given how new this feature is, but it all worked!  Very cool.  It was fun to scroll from the beginning and watching the symbol table get populated as shards were completing and collectors were running.

**Pull request merged**